src/main/java/org/jenkinsci/plugins/vsphere/VSphereNodeReconcileWork.java

```diff
@@ -0,0 +1,91 @@
+package org.jenkinsci.plugins.vsphere;
+
+import hudson.Extension;
+import hudson.Functions;
+import hudson.model.TaskListener;
+import hudson.model.AsyncPeriodicWork;
+import hudson.slaves.AbstractCloudSlave;
+import hudson.slaves.Cloud;
+import hudson.model.Label;
+import jenkins.model.Jenkins;
+import org.jenkinsci.plugins.vSphereCloud;
+import org.jenkinsci.plugins.vSphereCloudSlaveComputer;
+import org.jenkinsci.plugins.vSphereCloudSlaveTemplate;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * A {@link AsyncPeriodicWork} that reconciles nodes to meet template values.
+ * <p>
+ * The async work will check the number of deployed nodes and provision (or
+ * delete) additional ones to meet template values. The check is happening every
+ * 2 minutes.
+ */
+@Extension
+@Restricted(NoExternalUse.class)
+public final class VSphereNodeReconcileWork extends AsyncPeriodicWork {
+    private static final Logger LOGGER = Logger.getLogger(VSphereNodeReconcileWork.class.getName());
+
+    public VSphereNodeReconcileWork() {
+        super("Vsphere nodes reconciliation");
+    }
+
+    @Override
+    public long getRecurrencePeriod() {
+        return Functions.getIsUnitTest() ? Long.MAX_VALUE : MIN * 2;
+    }
+
+    @Override
+    public void execute(TaskListener listener) {
+        for (Cloud cloud : Jenkins.getActiveInstance().clouds) {
+            if (!(cloud instanceof vSphereCloud)) continue;
+            for (vSphereCloudSlaveTemplate template : ((vSphereCloud) cloud).getTemplates()) {
+                String templateLabel = template.getLabelString();
+                Label label = Label.get(templateLabel);
+
+                int instancesMin = template.getInstancesMin();
+                List<vSphereCloudSlaveComputer> idleNodes = template.getIdleNodes();
+                List<vSphereCloudSlaveComputer> runningNodes = template.getOnlineNodes();
+                // Get max number of nodes that could be provisioned
+                int globalMaxNodes = ((vSphereCloud) cloud).getInstanceCap();
+                int templateMaxNodes = template.getTemplateInstanceCap();
+                int maxNodes = Math.min(globalMaxNodes, templateMaxNodes);
+
+                // if maxNumber is lower than instancesMin, we have to ignore instancesMin
+                int toProvision = Math.min(instancesMin - idleNodes.size(), maxNodes - runningNodes.size());
```

Comment:
I'm not convinced that it's as simple as this as there's a delay between when a provisioning is started and when Jenkins is given a new agent and hence creates a new `vSphereCloudSlaveComputer` instance (and again when it's time to delete a VM too).

If you check the logic in vSphereCloud.provision(...) you'll see that it calls `ensureLists()` and then interrogates its `templateState` data to count how many VMs exist.
Some of those VMs will be in a state where the vSphere service has been asked to create a new VM but Jenkins won't know it exists yet; some of them will be fully created VMs that aren't online yet; some will be online and either busy or idle; some will be in the process of being deleted.
That `templateState` data structure is the "one true source of information" that we have (in memory) of what VMs we've asked for, what we've got, and what we've asked to delete.

The logic in use here only considers VMs that are fully created and online (and the subset of those which are idle); it does not consider those "not yet online" or "those being deleted".

I think that what we need here is to use the same logic (or at least be consistent with it) as used in vSphereCloud so that VMs which are being created are counted against the total present in the cloud, otherwise we could easily overshoot the desired `instancesMin` value (e.g. if VMs take so long to create that `public void execute(TaskListener listener)` gets called again before instances it's already requested have "arrived".
IMO this "pre-create" functionality will be most valued by folks whose vSphere infrastructure is slow to create new VMs, so this code _must_ cope with the infrastructure being slow.